### PR TITLE
add progressListener for android when using FormData to upload files

### PR DESCRIPTION
**Pull request by TanTan-TT**
When using FormData upload images or files, in Android version, network module cannot send an event for showing progress.
This PR will solve this issue.
I changed example in XHRExample for Android, you can see uploading progress in warning yellow bar.

### CODE REVIEW

**Comment by facebook-github-bot-7** [comment]
By analyzing the blame information on this pull request, we identified @astreet and @lexs to be potential reviewers.

**Comment by facebook-github-bot-5** [comment]
@TanTan-TT updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@TanTan-TT updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@TanTan-TT updated the pull request.

**Comment by TanTan-TT** [comment]
@satya164 Could you review this? CI error is unrelated.

**Comment by facebook-github-bot-6** [comment]
@TanTan-TT updated the pull request.

File: Examples/UIExplorer/XHRExample.android.js [hunk 1]
```diff
@@ -201,9 +201,10 @@ class FormUploader extends React.Component {
     this.state.textParams.forEach(
       (param) => formdata.append(param.name, param.value)
     );
+
     if (xhr.upload) {
       xhr.upload.onprogress = (event) => {
-        console.log('upload onprogress', event);
+        console.warn('upload onprogress', event);
```
Comment (satya164) [inline Examples/UIExplorer/XHRExample.android.js]: can we remove this log?

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 2]
```diff
@@ -233,7 +233,19 @@ public void sendRequest(
       if (multipartBuilder == null) {
         return;
       }
-      requestBuilder.method(method, multipartBuilder.build());
+
+      requestBuilder.method(method, RequestBodyUtil.createProgressRequest(multipartBuilder.build(),new ProgressRequestListener(){
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: nit: space after `multipartBuilder.build(),`

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 3]
```diff
@@ -290,6 +302,18 @@ private void readWithProgress(
...
   }
 
+  private static boolean shouldDispatch(long now, long last) {
+    return last + CHUNK_TIMEOUT_NS < now;
+  }
+
+  private void onDataSend(ExecutorToken ExecutorToken, int requestId, long progress, long total) {
+    WritableArray args = Arguments.createArray();
+    args.pushInt(requestId);
+    args.pushInt((int)progress);
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: nit: space after `(int)`

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/RequestBodyUtil.java [hunk 4]
```diff
@@ -115,6 +115,13 @@ public void writeTo(BufferedSink sink) throws IOException {
   }
 
   /**
+   * Creates a ProgressRequestBody that can be used for showing uploading progress
+   */
+  public static ProgressRequestBody createProgressRequest(RequestBody requestBody, ProgressRequestListener listener){
+    return new ProgressRequestBody(requestBody,listener);
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/RequestBodyUtil.java]: nit: space after `requestBody,`

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/RequestBodyUtil.java [hunk 5]
```diff
@@ -115,6 +115,13 @@ public void writeTo(BufferedSink sink) throws IOException {
   }
 
   /**
+   * Creates a ProgressRequestBody that can be used for showing uploading progress
+   */
+  public static ProgressRequestBody createProgressRequest(RequestBody requestBody, ProgressRequestListener listener){
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/RequestBodyUtil.java]: nit: space before `{`

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/ProgressRequestBody.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import com.squareup.okhttp.internal.Util;
+import okio.BufferedSink;
+import okio.Buffer;
+import okio.Sink;
+import okio.ForwardingSink;
+import okio.ByteString;
+import okio.Okio;
+import okio.Source;
+
+
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/ProgressRequestBody.java]: nit: remove extra new line

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/ProgressRequestBody.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+public class ProgressRequestBody extends RequestBody {
+
+  private final RequestBody requestBody;
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/ProgressRequestBody.java]: nit: name the variables as `mRequestBody` and `mProgressListener`

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/ProgressRequestBody.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  @Override
+  public void writeTo(BufferedSink sink) throws IOException {
+      if (bufferedSink == null) {
+          bufferedSink = Okio.buffer(sink(sink));
+      }
+
+      requestBody.writeTo(bufferedSink);
+
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/ProgressRequestBody.java]: nit: remove extra new lines

File: Libraries/Network/XMLHttpRequest.android.js [hunk 9]
```diff
@@ -26,6 +26,11 @@ function convertHeadersMapToArray(headers: Object): Array<Header> {
 }
 
 class XMLHttpRequest extends XMLHttpRequestBase {
+  constructor() {
+    super();
+    // android supports upload
+    this.upload = {};
```
Comment (satya164) [inline Libraries/Network/XMLHttpRequest.android.js]: we should move this to `XMLHttpRequestBase` as both Android and iOS support upload

File: Libraries/Network/XMLHttpRequest.android.js [hunk 10]
```diff
@@ -26,6 +26,11 @@ function convertHeadersMapToArray(headers: Object): Array<Header> {
...
+  }
```
Comment (satya164) [inline Libraries/Network/XMLHttpRequest.android.js]: nit: add new line after `}`

File: Examples/UIExplorer/XHRExample.android.js [hunk 11]
```diff
@@ -201,9 +201,10 @@ class FormUploader extends React.Component {
     this.state.textParams.forEach(
       (param) => formdata.append(param.name, param.value)
     );
+
     if (xhr.upload) {
```
Comment (satya164) [inline Examples/UIExplorer/XHRExample.android.js]: I think we can remove this check now

**Comment by facebook-github-bot-5** [comment]
@TanTan-TT updated the pull request.

**Comment by TanTan-TT** [comment]
@satya164 I hope this can be landed, please check it

**Comment by TanTan-TT** [comment]
@mkonicek Could you arrange someone review this code please? I hope this can be landed in.

**Comment by satya164** [comment]
@TanTan-TT sorry, on vacation right now. Will review as soon as I get back.

**Comment by mkonicek** [comment]
👍 Have a nice vacation Satyajit.

**Comment by TanTan-TT** [comment]
@satya164 sorry for bothering you again? R u still on vocation?

File: Libraries/Network/XMLHttpRequest.android.js [hunk 12]
```diff
@@ -26,6 +26,9 @@ function convertHeadersMapToArray(headers: Object): Array<Header> {
 }
 
 class XMLHttpRequest extends XMLHttpRequestBase {
+  constructor() {
+    super();
```
Comment (satya164) [inline Libraries/Network/XMLHttpRequest.android.js]: this shouldn't be needed

File: Libraries/Network/XMLHttpRequest.ios.js [hunk 13]
```diff
@@ -19,10 +19,7 @@ var XMLHttpRequestBase = require('XMLHttpRequestBase');
 class XMLHttpRequest extends XMLHttpRequestBase {
   constructor() {
     super();
-    // iOS supports upload
-    this.upload = {};
```
Comment (satya164) [inline Libraries/Network/XMLHttpRequest.ios.js]: we could remove the constructor entirely.

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 14]
```diff
@@ -233,7 +233,19 @@ public void sendRequest(
       if (multipartBuilder == null) {
         return;
       }
-      requestBuilder.method(method, multipartBuilder.build());
+
+      requestBuilder.method(method, RequestBodyUtil.createProgressRequest(multipartBuilder.build(), new ProgressRequestListener(){
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: nit: space before `{`

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/ProgressRequestBody.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+import okio.Source;
+
+public class ProgressRequestBody extends RequestBody {
+
+  private final RequestBody mRequestBody;
+  private final ProgressRequestListener mProgressListener;
+  private BufferedSink mBufferedSink;
+
+  public ProgressRequestBody(RequestBody requestBody, ProgressRequestListener progressListener) {
+      this.mRequestBody = requestBody;
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/ProgressRequestBody.java]: unnecessary `this`

File: ReactAndroid/src/test/java/com/facebook/react/modules/network/NetworkingModuleTest.java [hunk 16]
```diff
@@ -265,6 +267,7 @@ public void testMultipartPostRequestSimple() throws Exception {
         .thenReturn(mock(InputStream.class));
     when(RequestBodyUtil.create(any(MediaType.class), any(InputStream.class)))
         .thenReturn(mock(RequestBody.class));
+    when(RequestBodyUtil.createProgressRequest(any(RequestBody.class), any(ProgressRequestListener.class))).thenCallRealMethod();
```
Comment (satya164) [inline ReactAndroid/src/test/java/com/facebook/react/modules/network/NetworkingModuleTest.java]: 👍

File: Examples/UIExplorer/XHRExample.android.js [hunk 17]
```diff
@@ -201,19 +226,29 @@ class FormUploader extends React.Component {
...
     );
-    if (xhr.upload) {
-      xhr.upload.onprogress = (event) => {
-        console.log('upload onprogress', event);
-        if (event.lengthComputable) {
-          this.setState({uploadProgress: event.loaded / event.total});
-        }
-      };
+    if (this.state.randomPhoto) {
+      formdata.append('image', {...this.state.randomPhoto,type:'image/jpg', name: 'image.jpg'});
```
Comment (satya164) [inline Examples/UIExplorer/XHRExample.android.js]: nit: space after comma

File: Examples/UIExplorer/XHRExample.ios.js [hunk 18]
```diff
@@ -224,14 +224,13 @@ class FormUploader extends React.Component {
...
     );
-    if (xhr.upload) {
-      xhr.upload.onprogress = (event) => {
-        console.log('upload onprogress', event);
-        if (event.lengthComputable) {
-          this.setState({uploadProgress: event.loaded / event.total});
-        }
-      };
-    }
+    xhr.upload.onprogress = (event) => {
```
Comment (satya164) [inline Examples/UIExplorer/XHRExample.ios.js]: we could probably merge the iOS and Android examples. What's the difference?

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 19]
```diff
@@ -233,7 +233,19 @@ public void sendRequest(
...
+        long last = System.nanoTime();
+
+        @Override
+        public void onRequestProgress(long bytesWritten, long contentLength, boolean done){
+          long now = System.nanoTime();
+          if (done || shouldDispatch(now, last)) {
+            onDataSend(executorToken,requestId,bytesWritten,contentLength);
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: nit: space after comma

**Comment by satya164** [comment]
Sorry for the late and thanks for the ping. I'm not super-familiar with this code, but looks good overall. Looks like there are conflicts with master, can you address the comments and rebase?

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java (hunk 14 above)
Comment (Frosty92) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: I am getting the following error when I try to run:

```
error: cannot find symbol
      @Nullable List<NetworkInterceptorCreator> networkInterceptorCreators) {
                     ^
  symbol:   class NetworkInterceptorCreator
  location: class NetworkingModule

```

**Comment by facebook-github-bot-1** [comment]
@TanTan-TT updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@TanTan-TT updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@TanTan-TT updated the pull request.

**Comment by TanTan-TT** [comment]
@satya164 Error is unrelated. And "fb:no-direct-merges " is still in yellow, I don't know why. Do you think it can be landed in?

File: Examples/UIExplorer/XHRExample.ios.js (hunk 18 above)
Comment (TanTan-TT) [inline Examples/UIExplorer/XHRExample.ios.js]: I think in different files can be easy to maintain.

**Comment by satya164** [comment]
Yeah, build seems to be broken on master. The no-direct-merges will always stay as pending to prevent accidental direct merging. We use the facebook bot to merge instead.

Can you address my other comments?

**Comment by TanTan-TT** [comment]
@satya164 Sorry, I didn't understand your comment. What do you mean "Can you address my other comments?"? I should have fix all your reviews.

File: Examples/UIExplorer/XHRExample.android.js (hunk 17 above)
Comment (satya164) [inline Examples/UIExplorer/XHRExample.android.js]: can you fix this nit also?

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 20]
```diff
@@ -56,7 +56,7 @@
   private static final String REQUEST_BODY_KEY_URI = "uri";
   private static final String REQUEST_BODY_KEY_FORMDATA = "formData";
   private static final String USER_AGENT_HEADER_NAME = "user-agent";
-
+  private static final int CHUNK_TIMEOUT_NS = 100 * 1000000; // 100ms
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: What does `NS` mean here?

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/ProgressRequestListener.java [hunk 21]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+package com.facebook.react.modules.network;
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/ProgressRequestListener.java]: nit: add newline before this line

**Comment by satya164** [comment]
@TanTan-TT Just the few nits I commented on and then we shall merge.

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java (hunk 20 above)
Comment (TanTan-TT) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: @satya164 This is the old code, but in previous version, this code was deleted. Before that version, this code is used in downloading scenario. I just reuse this code. Thus I don't know what ```NS``` really mean

**Comment by facebook-github-bot-6** [comment]
@TanTan-TT updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@TanTan-TT updated the pull request.

**Comment by TanTan-TT** [comment]
@satya164 All fixed

**Comment by satya164** [comment]
This one is still left - https://github.com/facebook/react-native/pull/7256#discussion_r64987786

**Comment by TanTan-TT** [comment]
@satya164 Sorry for missing this. I fixed it.

**Comment by facebook-github-bot-9** [comment]
@TanTan-TT updated the pull request.

**Comment by satya164** [comment]
Thanks. I'll shipit when tests pass.

**Comment by TanTan-TT** [comment]
@satya164 Thank you so much.

**Comment by satya164** [comment]
@facebook-github-bot shipit

**Comment by TanTan-TT** [comment]
@satya164 Do I need to close this PR by myself or bot will do this?

**Comment by TanTan-TT** [comment]
@satya164 Sorry for bothering again. It seems bot had not run up. Could you check it again?

**Comment by satya164** [comment]
@TanTan-TT You won't have to do anything.

@facebook-github-bot shipit

**Comment by facebook-github-bot-4** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/489984071209836/int_phab) to review.

**Pull request closed without merging**